### DISCUSSION TITLE
Timecode per/ word a future option?

### DISCUSSION THREAD

**Discussion by samelie** [openai/whisper#26] [Q&A] [answered]
Right now we get a range printed , eg: `[00:00.000 --> 00:05.720]`. Would be great to have a way to print out the timecode next to the word. This would unlock lots of interactive behavior.
Thank you for releasing this achievement

**Comment by RaulKite** [comment]
It will be a killer option to use  in may many cases

**Reply by kospl** [reply]
name one use case for this feature

**Reply by DAVIDSystems** [reply]
audioediting with text, subtitling with word highlighting and many more.

**Reply by DAVIDSystems** [reply]
![image](https://user-images.githubusercontent.com/40237707/192154487-08680041-48ff-4472-9f18-d8e41f59cfb1.png)

**Comment by jongwook** [comment, marked as answer]
(Duplicate of #3) Getting word-level timestamps are not directly supported, but it could be possible using the predicted distribution over the timestamp tokens or the cross-attention weights.

Currently, the predicted timestamps tend to be biased towards integers, and there are some failure modes where the timestamps can be constantly shifted, making reliable word-level timestamp prediction difficult. Once this is solved by us or the community, I agree that it'd be a great addition to this repo.

**Reply by melindadevins** [reply]
When transcript is displayed on video, the word being spoken is often highlighted. The word level timestamp makes it possible. Really appreciate it if the timestamp can be obtained.

**Reply by jayn1985** [reply]
word level timestamp is really helpful in many cases, hope it can be supported in the later version.

**Comment by edgartaor** [comment]
Could something like this work?

https://nvidia.github.io/OpenSeq2Seq/html/speech-recognition/speech-to-text-align.html

**Comment by sandalwoodsh** [comment]
I agree that such a feature would be tremendously awesome!

**Comment by jongwook** [comment]
I've made a demo of obtaining word-level timestamps using the cross-attention patterns in [the multilingual ASR notebook](https://github.com/openai/whisper/blob/main/notebooks/Multilingual_ASR.ipynb)

![image](https://user-images.githubusercontent.com/266841/210467619-4384e3d0-522b-4e84-8c64-60e6422eed26.png)

**Reply by octimot** [reply]
@jongwook Thanks for this update! It seems to work pretty well, but I'm having trouble scaling it up to audio longer than 30 seconds:

The call `whisper_model(mel.unsqueeze(0), tokens.unsqueeze(0))` triggers `IndexError: index out of range in self` in 
`return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)`

What would be the best approach to make it take each 30sec window at a time? Would you mind giving a rough example?

Cheers!

**Reply by RaulKite** [reply]
Have you tried this repo? 

https://github.com/linto-ai/whisper-timestamped

It is based in @jongwook notebook and have advanced in that way

**Reply by octimot** [reply]
@RaulKite 
Yes, thank you! 

I believe whisper-timestamped is doing the word-alignment on the fly which is cool, but I'd like to understand how to scale the basic implementation that @jongwook proposed, just to have a bit more flexibility.
